ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingSwitchingPeerTask.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright contributors to Hyperledger Besu
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.eth.manager.task;
+
+import static org.hyperledger.besu.util.Slf4jLambdaHelper.traceLambda;
+
+import org.hyperledger.besu.ethereum.eth.manager.EthContext;
+import org.hyperledger.besu.ethereum.eth.manager.EthPeer;
+import org.hyperledger.besu.ethereum.eth.manager.exceptions.NoAvailablePeersException;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Predicate;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractRetryingSwitchingPeerTask<T> extends AbstractRetryingPeerTask<T> {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(AbstractRetryingSwitchingPeerTask.class);
+
+  private final Set<EthPeer> triedPeers = new HashSet<>();
+
+  protected AbstractRetryingSwitchingPeerTask(
+      final EthContext ethContext,
+      final MetricsSystem metricsSystem,
+      final Predicate<T> isEmptyResponse,
+      final int maxRetries) {
+    super(ethContext, maxRetries, isEmptyResponse, metricsSystem);
+  }
+
+  @Override
+  public void assignPeer(final EthPeer peer) {
+    super.assignPeer(peer);
+    triedPeers.add(peer);
+  }
+
+  protected abstract CompletableFuture<T> executeTaskOnCurrentPeer(final EthPeer peer);
+
+  @Override
+  protected CompletableFuture<T> executePeerTask(final Optional<EthPeer> assignedPeer) {
+
+    final Optional<EthPeer> maybePeer =
+        assignedPeer
+            .filter(u -> getRetryCount() == 1) // first try with the assigned peer if present
+            .map(Optional::of)
+            .orElseGet(this::selectNextPeer); // otherwise select a new one from the pool
+
+    if (maybePeer.isEmpty()) {
+      traceLambda(
+          LOG,
+          "No peer found to try to execute task at attempt {}, tried peers {}",
+          triedPeers::toString,
```

Comment:
this param and next one should be swapped

Reply:
done